lib/common/patchset.c

```diff
@@ -581,35 +591,40 @@ xml_log_patchset(uint8_t log_level, const char *function, xmlNode *patchset)
                 }
             }
         }
-        return;
-    }
 
-    if (log_level < LOG_DEBUG) {
-        options |= xml_log_option_diff_short;
-    }
+    } else {
+        int options = xml_log_option_formatted;
+        const xmlNode *removed = NULL;
+        const xmlNode *added = NULL;
+        bool is_first = true;
 
-    removed = find_xml_node(patchset, "diff-removed", FALSE);
-    for (child = pcmk__xml_first_child(removed); child != NULL;
-         child = pcmk__xml_next(child)) {
-        log_data_element(log_level, __FILE__, __func__, __LINE__, "- ", child,
-                         0, options|xml_log_option_diff_minus);
-        if (is_first) {
-            is_first = FALSE;
-        } else {
-            do_crm_log(log_level, " --- ");
+        if (log_level < LOG_DEBUG) {
+            options |= xml_log_option_diff_short;
         }
-    }
 
-    is_first = TRUE;
-    added = find_xml_node(patchset, "diff-added", FALSE);
-    for (child = pcmk__xml_first_child(added); child != NULL;
-         child = pcmk__xml_next(child)) {
-        log_data_element(log_level, __FILE__, __func__, __LINE__, "+ ", child,
-                         0, options|xml_log_option_diff_plus);
-        if (is_first) {
-            is_first = FALSE;
-        } else {
-            do_crm_log(log_level, " +++ ");
+        removed = find_xml_node(patchset, "diff-removed", FALSE);
+        for (child = pcmk__xml_first_child(removed); child != NULL;
+             child = pcmk__xml_next(child)) {
+            log_data_element(log_level, __FILE__, __func__, __LINE__, "- ",
+                             child, 0, options|xml_log_option_diff_minus);
+            if (is_first) {
+                is_first = false;
+            } else {
+                do_crm_log(log_level, " --- ");
```

Comment:
Not really sure why the `---` and `+++` lines come **after** `log_data_element()` and aren't logged on the first iteration... are these a weird-looking divider?

Reply:
Let's make a note in the change log that "11" was a typo all along.

Reply:
This also means that when we do formatted output for `crm_shadow`/`crm_diff`, v1 patchsets **won't** need to validate against the schema, right? I wrote a v2 schema a couple of weeks ago. I was dreading the v1 if it turned out to be necessary.

Reply:
Let's put the patch in CDATA and not bother validating it in any case. It's effectively external XML passed along by the tool rather than tool output itself, much like querying the CIB.

Reply:
In fact that's what ->output_xml() does IIRC

Reply:
We can definitely do that -- a v2 diff does have a particular format though: https://gist.github.com/nrwahl2/83ae39b63ce0f023713fc8589b1b08db

I didn't do any validation of the CIB pieces that are contained inside the diff, only of the diff structure itself. I figure having a schema also makes it easier for API users to understand what they're getting.

Reply:
Just like the CIB schema is separate, if we want a patchset schema, it should have its own. The user would validate a patchset itself separately from validating the tool output that produced it.

Reply:
Makes sense